pkg/api/rest/update.go

```diff
@@ -37,9 +37,16 @@ type RESTUpdateStrategy interface {
 	// the object.  For example: remove fields that are not to be persisted,
 	// sort order-insensitive list fields, etc.
 	PrepareForUpdate(obj, old runtime.Object)
-	// ValidateUpdate is invoked after default fields in the object have been filled in before
-	// the object is persisted.
+	// ValidateUpdate is invoked after default fields in the object have
+	// been filled in, but before the object is persisted.  This method is
+	// responsible for ensuring that any disallowed changes are flagged.
+	// For example: changing immutable fields.  It does not need to
+	// re-validate the whole object - that is what Validate() is for.
 	ValidateUpdate(ctx api.Context, obj, old runtime.Object) fielderrors.ValidationErrorList
+	// Validate is invoked after default fields in the object have been
+	// filled in and after the update itself has been validated, but before
+	// the object is persisted.
+	Validate(ctx api.Context, obj runtime.Object) fielderrors.ValidationErrorList
```

Comment:
> On Mar 31, 2015, at 11:27 PM, Tim Hockin <notifications@github.com> wrote:
> 
> In pkg/api/rest/update.go:
> 
> >  	ValidateUpdate(ctx api.Context, obj, old runtime.Object) fielderrors.ValidationErrorList
> > +	// Validate is invoked after default fields in the object have been
> > +	// filled in and after the update itself has been validated, but before
> > +	// the object is persisted.
> > +	Validate(ctx api.Context, obj runtime.Object) fielderrors.ValidationErrorList
> That's fine in theory, but is the distinction real? Is there REALLY anything that will go in ValidateCreate() that is not also needed in ValidateUpdate(), and there for should be in the common ValidateObject()? Can we find an actual example?

I mean that validatecreate() is what the RestCreateStrategy method should be called.  I think we agree that ValidateUpdate on the strategy for each type should "fully validate" the object, which usually will be calling VU and V together.

> Already this has a flaw that errors in Validate() will not be caught until ValidateUpdate passes. And some things get double validated (metadata). The real alternative is to punt down to each strategy and just tell them "thou shalt validate object when validating updates". Already our validation is too loose IMO...

That's the alternative I'm arguing for.

> …
> —
> Reply to this email directly or view it on GitHub.
>